Gamma optimizations and benchmarks

eightysteele opened the pull request:
Turns out that pre-computing constants and type-hinting functions has a big performance gain:

- `(gamma 2.5)` 77.40% speedup (execution mean 4.889986 µs → 1.104972 µs)
- `(log-gamma 10)` 63.92% speedup (execution mean 4.845715 µs → 1.748155 µs)
- `(log-gamma-1p -0.05)` 76.86% speedup (execution mean 4.981068 µs → 1.152443 µs)
- `(inv-gamma-1pm1 -0.05)` 77.85% speedup (execution mean 4.582663 µs → 1.015241 µs)
- `(lanczos-approximation 170)` 73.15% speedup (execution 4.276010 µs → 1.148002 µs)

Did this for `math.gamma` and `math.utils`. Also experimenting with a way to define and run benchmarks like tests: Benchmarks are defined in the `/benchmark` directory, then functions to benchmark are tagged with `^:benchmark` and profiled with Criterium. Just run `bb benchmark:clj` to compute benchmarks! See `gamma_benchmark.cljc` for an example.

sritchie commented on benchmark/benchmark_runner.cljc:
```diff
@@ -0,0 +1,48 @@
+  (->> (ns-publics (find-ns (symbol namespace)))
+       (keep (fn [[sym var]] (when (:benchmark (meta var)) sym)))))
+
+(defn run-benchmarks
```
I think you can do this with our normal test runner, without adding this whole namespace... https://github.com/cognitect-labs/test-runner see the include and exclude options: https://github.com/cognitect-labs/test-runner?tab=readme-ov-file#invoke-with-clojure--m-clojuremain

eightysteele replied:
Yeah, good call. I swapped in test runner—much better.

sritchie commented on benchmark/gen/distribution/math/gamma_benchmark.cljc:
```diff
@@ -0,0 +1,59 @@
+  (:require [gen.distribution.math.gamma :as g]
+            [criterium.core :as crit]))
+
+(defn ^:benchmark lanczos-approximation-computation-time
```
you should be able to use `deftest` and then the test selectors I mentioned above

sritchie commented on benchmark/gen/distribution/math/gamma_benchmark.cljc:
```diff
@@ -0,0 +1,59 @@
+            (crit/quick-bench
+                (dotimes [_ n]
+                  (g/inv-gamma-1pm1 x))))]
+    (mapv f x)))
```
for side effects, use `run!` instead of `mapv`

eightysteele replied:
Nice, done.

sritchie commented on src/gen/distribution/math/utils.cljc:
```diff
@@ -1,14 +1,14 @@
   (Math/log (* 2 Math/PI)))
 
-(def ^:no-doc sqrt-2pi
+(def ^:no-doc ^:const sqrt-2pi
```
wild, I have some annotations I need to add in Emmy!

eightysteele replied:
So good!

eightysteele commented:
@sritchie I addressed the comments and benchmark now ride on test runner, boom. 

I'm assuming that when a benchmark test runs, we want it to fail if it diverged within some threshold of a baseline. To define a baseline, it seems reasonable to have something like this:

```clojure
(def ^:no-doc ^:const lanczos-approximation-baseline
  [{:n 15
    :x 9
    :mean 3.127368933269932E-6
    :variance 1.5780434242573877E-14
    :lower-q 2.984515105404977E-6
    :upper-q 3.320172412426692E-6}
   {:n 15
    :x 100
    :mean 2.0160584455226908E-7
    :variance 7.928671640576774E-18
    :lower-q 1.9907849968797706E-7
    :upper-q 2.0489675953499433E-7}
   {:n 15
    :x 170
    :mean 2.040524636004644E-7
    :variance 5.795520133553684E-18
    :lower-q 2.0120967339244923E-7
    :upper-q 2.0719692107157572E-7}])
```

and then the actual benchmark test just maps over entries in the baseline and runs `crit/quick-benchmark`, like so:

```clojure
(deftest ^:benchmark lanczos-approximation-benchmark
  []
  (let [f (fn[b]
            (let [n (:n b)
                  x (:x b)
                  mean (:mean b)
                  variance (:variance b)
                  lower-q (:lower-q b)
                  upper-q (:upper-q b)
                  stats (crit/quick-benchmark
                         (dotimes [_ n]
                           (g/lanczos-approximation x)) {})]
              (testing "against lanczos approximation baseline"
                (with-comparator (within 1e-3)
                  (is (ish? mean (first (:mean stats))) x)
                  (is (ish? variance (first (:variance stats))) x)
                  (is (ish? lower-q (first (:lower-q stats))) x)
                  (is (ish? upper-q (first (:upper-q stats))) x)))))]
    (run! f lanczos-approximation-baseline)))
    ```
    
    This is implemented right now, so you can run `bb benchmark:clj` to see it work.

sritchie commented on benchmark/gen/distribution/math/gamma_benchmark.cljc:
```diff
@@ -0,0 +1,91 @@
+        x [0.1 1.5 2.5 8.0 10.0]
+        f (fn[x]
+            (println "log-gamma: " x)
+            (crit/quick-bench
```
bummer that criterium can't take a label!

sritchie commented on deps.edn:
```diff
@@ -33,6 +33,15 @@
+                {:git/tag "v0.5.1" :git/sha "dfb30dd"}
+                criterium/criterium {:mvn/version "0.4.6"}}
+   :main-opts ["-m" "cognitect.test-runner"]
+   :exec-args {:includes [:benchmark] :dirs ["benchmark"] :patterns [".*-benchmark$"]}
```
okay I see here that we specify the test name pattern... but what does the metadata do?

eightysteele replied:
yeah good q! so from test-runner, `:patterns` is for the test filenames to run and `:includes` is the collection of test metadata keywords to include. so, this says "find all files with a name that ends with "benchmark" and in those files run any function with the `:benchmark` metadata.

sritchie reviewed (approved):
one comment...

eightysteele commented:
@sritchie yep, i responded—lmk if that explanation about `:benchmark` metadata makes sense!

eightysteele commented:
💥
